### ISSUE TITLE
Email Validation Hangs at 100% CPU usage

### ISSUE THREAD

**Issue by arthurbelair** [colinhacks/zod#2580]
Hello,

I came across an issue with the email validation for a specific input string (a long and complex email address).
When attempting to validate/parse the string, the process immediately goes to 100% CPU usage, and never actually finishes parsing. I have waited for ~15 minutes max.
 
Here's some code to reproduce, which also shows 2 other RegExes for which the problem doesn't occur:

```typescript
import {z} from "zod"

const email = "noreply1@1-90y790o645vz81zq6287dhb9x1x28ab4q521nuv7b1og4y9c3q.5r-h8uter0.sc016.case.sandbox.salesforce.com"

const otherRegex1 = /[a-z0-9!#$%&'*+/=?^_`{|}~-]+(?:\.[a-z0-9!#$%&'*+/=?^_`{|}~-]+)*@(?:[a-z0-9](?:[a-z0-9-]*[a-z0-9])?\.)+[a-z0-9](?:[a-z0-9-]*[a-z0-9])?/gi
const otherRegex2 = /^[\w-\.]+@([\w-]+\.)+[\w-]{2,4}$/gi

console.log(otherRegex1.test(email)) // true
console.log(otherRegex2.test(email)) // true

const parsed = z.string().email().parse(email) // hangs at 100% CPU

console.log(parsed) // never reached
``` 

![Screenshot 2023-07-14 at 10 58 23 AM](https://github.com/colinhacks/zod/assets/13756082/63d01f8a-4f6c-44c7-93c9-63a7083541b8)

Using Zod `v3.21.4`
on a 2023 M2 Macbook Pro

**Comment by ramonfallon** [comment]
Here is another example
```
 youdonothavemypermission.tostoremyinfo.com
```

**Comment by callmeberzerker** [comment]
```
const otherRegex1 = /[a-z0-9!#$%&'*+/=?^_`{|}~-]+(?:\.[a-z0-9!#$%&'*+/=?^_`{|}~-]+)*@(?:[a-z0-9](?:[a-z0-9-]*[a-z0-9])?\.)+[a-z0-9](?:[a-z0-9-]*[a-z0-9])?/gi
const otherRegex2 = /^[\w-\.]+@([\w-]+\.)+[\w-]{2,4}$/gi
```

This is definitely a big issue. Can we replace the email regex with one of those 2 which are described in the first post?

**Comment by colinhacks** [comment]
Fixed by https://github.com/colinhacks/zod/pull/2824

Landed in Zod v3.22.3
